WT-5766 Separate out internal and shared transaction data

tetsuo-cpp opened the pull request:
As part of reading a history store modify, we move our transaction's read timestamp forward to make intermediate modifies and the eventual base update visible. During this process, the oldest timestamp can be moved forward and be placed ahead of the original read timestamp.

When we attempt to move the read timestamp back, we hit an assertion. And in release mode, we probably have visibility issues.

To fix this, I'm introducing a `WT_TXN::pinned_read_timestamp` which is used to determine how far we can move the oldest. So all the logic where we iterate the read queue to check the minimum read timestamp will be using `pinned_read_timestamp`. Any actual visibility checks in the active transaction's thread will be using the `read_timestamp` which is the member that we'll be mutating in our history store logic.

This is just a first cut to understand what this change will look like. I still need to work on getting a repro before merging this.

tetsuo-cpp commented on src/include/txn.h:
```diff
@@ -288,6 +288,13 @@ struct __wt_txn {
+     * mechanism, we can move the read timestamp forward so we need to keep track of the original
+     * read timestamp to know what history should be pinned in front of oldest.
+     */
+    wt_timestamp_t first_read_timestamp;
```
The reason I didn't modify the querying is because `read_timestamp != first_read_timestamp` for only a small window while we're doing a history store modify read before getting restored.

So we'll never be able to query the session at a time where they're not equal.

keithbostic replied:
> So we'll never be able to query the session at a time where they're not equal.

I get that, but I think of it as future-proofing. This new field is the "truth" for everything but the current search, right? If that statement is correct, that should be reflected in the code?

tetsuo-cpp replied:
487e5f3

tetsuo-cpp commented on src/include/txn.h:
```diff
@@ -288,6 +288,13 @@ struct __wt_txn {
+     * mechanism, we can move the read timestamp forward so we need to keep track of the original
+     * read timestamp to know what history should be pinned in front of oldest.
+     */
+    wt_timestamp_t first_read_timestamp;
```
8b85cd5

tetsuo-cpp commented on src/include/txn.h:
```diff
@@ -288,6 +288,13 @@ struct __wt_txn {
+     * mechanism, we can move the read timestamp forward so we need to keep track of the original
+     * read timestamp to know what history should be pinned in front of oldest.
+     */
+    wt_timestamp_t first_read_timestamp;
```
43e898e

tetsuo-cpp commented:
Ok, I've addressed all comments.
The following commits are to explore putting this information into a `WT_TXN_SHARED` structure and only reading the `WT_TXN` from the owning session's thread. If this gets too involved, we can always revert back to 43e898e.

tetsuo-cpp commented on src/include/txn.h:
```diff
@@ -316,43 +345,27 @@ struct __wt_txn {
-#define WT_TXN_HAS_TS_COMMIT 0x000010u
-#define WT_TXN_HAS_TS_DURABLE 0x000020u
-#define WT_TXN_HAS_TS_PREPARE 0x000040u
-#define WT_TXN_HAS_TS_READ 0x000080u
```
Wasn't sure whether to move this one across to the shared flags since we have a `read_timestamp` here too.

tetsuo-cpp replied:
Yeah, my understanding is the same.

tetsuo-cpp commented on src/txn/txn_timestamp.c:
```diff
@@ -280,11 +287,11 @@ __txn_global_query_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *tsp, cons
         __wt_readlock(session, &txn_global->durable_timestamp_rwlock);
-        TAILQ_FOREACH (txn, &txn_global->durable_timestamph, durable_timestampq) {
-            if (txn->clear_durable_q)
+        TAILQ_FOREACH (txn_state, &txn_global->durable_timestamph, durable_timestampq) {
```
It's actually possible to iterate over shared transactions by traversing `txn_global->states`. However, I think in some cases the order of these queues is significant so this is a non-trivial change. For now, I've just made the queues contain `WT_TXN_SHARE` instead of `WT_TXN`.

keithbostic commented on src/txn/txn.c:
```diff
@@ -1402,8 +1419,8 @@ __wt_txn_init(WT_SESSION_IMPL *session, WT_SESSION_IMPL *session_ret)
-        WT_TXN_STATE *txn_state;
-        txn_state = WT_SESSION_TXN_STATE(session_ret);
+        WT_TXN_SHARED *txn_state;
+        txn_state = WT_SESSION_TXN_SHARED(session_ret);
```
If we write this as:
```
WT_ASSERT(session, S2C(session_ret)->txn_global.states == NULL ||
      WT_SESSION_TXN_SHARED(session_ret)->pinned_id == WT_TXN_NONE);
```
I don't think we need the #ifdef or local variable.

tetsuo-cpp replied:
c12daa2

keithbostic commented on src/txn/txn.c:
```diff
@@ -1727,17 +1744,19 @@ __wt_verbose_dump_txn_one(
-      txn->ckpt_lsn.l.offset, txn->full_ckpt ? "true" : "false",
+      __wt_timestamp_to_string(txn->read_timestamp, ts_string[4]),
+      __wt_timestamp_to_string(txn_state->pinned_read_timestamp, ts_string[5]),
+      txn->ckpt_lsn.l.file, txn->ckpt_lsn.l.offset, txn->full_ckpt ? "true" : "false",
```
We might want to group these by the structure they're in, but I don't feel strongly.

tetsuo-cpp replied:
2b4b2b9

keithbostic commented on src/txn/txn.c:
```diff
@@ -1822,7 +1841,8 @@ __wt_verbose_dump_txn(WT_SESSION_IMPL *session)
           s->pinned_id, s->metadata_pinned, sess->name == NULL ? "EMPTY" : sess->name));
-        WT_RET(__wt_verbose_dump_txn_one(session, &sess->txn, 0, NULL));
+        WT_RET(
+          __wt_verbose_dump_txn_one(session, &sess->txn, WT_SESSION_TXN_SHARED(sess), 0, NULL));
```
We could pass `sess` to `__wt_verbose_dump_txn_one` rather than passing in 2 arguments, both derived from `sess`?

tetsuo-cpp replied:
e391e70

tetsuo-cpp commented on src/include/txn.h:
```diff
@@ -99,13 +99,60 @@ typedef enum {
+    /* Set if need to clear from the durable queue */
+    bool clear_durable_q;
+
+    WT_TXN *internal_txn; /* A reference to the corresponding internal transaction data. */
```
5f9ebaa

Removed slop.

As for the `internal_txn` member, yeah we usually shouldn't be looking at that from other threads. The only exception (and this is why I've left this escape hatch here) is in `__txn_get_published_timestamp`. We need to compare the `commit_timestamp` with `first_commit_timestamp` to figure out whether we should return the `first_commit_timestamp` or in the case of a potentially resolved prepared transaction, `durable_timestamp`. We specifically have to check the timestamps because the transaction may be committed before getting removed from the queue and may have its flags cleared.

keithbostic commented on src/include/txn.i:
```diff
@@ -162,8 +162,10 @@ static inline void
 __txn_resolve_prepared_update(WT_SESSION_IMPL *session, WT_UPDATE *upd)
 {
     WT_TXN *txn;
+    WT_TXN_SHARED *txn_state;
```
The local variable is called `txn_state`, should it be `txn_shared`?

tetsuo-cpp replied:
I don't mind, it's still a state. Happy to do another rename though.

tetsuo-cpp replied:
Ok, I have a commit ready for this.
I'm not going to push it just yet as it will blow up the diffs and make it hard to review.

tetsuo-cpp replied:
d1462b8

keithbostic commented on src/txn/txn.c:
```diff
@@ -242,6 +242,13 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
+     * state. There are a few instances where we need to read the internal transaction data in a
+     * controlled manner so we need to leave an escape hatch here to allow this.
+     */
+    txn_state->internal_txn = txn;
```
Circling back to this field -- if we need this field, doesn't that mean the approach doesn't work?

tetsuo-cpp replied:
My point of view is just that the stuff in `WT_TXN_SHARED` can be read safely and easily without any caveats. Whereas reading directly from the `internal_txn` is a "do at your own risk" sort of thing and each usage of this member should be looked at carefully before going in.

The only usage at the moment is `__txn_get_published_timestamp` which I think is legitimate. Do you agree?

keithbostic replied:
@michaelcahill, would you please do a review?

michaelcahill replied:
I've pushed a change in 62e6121 that gets rid of `internal_txn` (by publishing a `pinned_durable_timestamp` separate from the `WT_TXN`-local `durable_timestamp`.  I think it's clearer this way: I was really struggling to convince myself that the previous code was correct...

tetsuo-cpp replied:
Thanks for the cleanup, @michaelcahill. I think I understand this.
I asked some questions on Slack but I'll record the responses here in case anyone has similar questions:

- The main change is to record the timestamp we use to insert into the durable queue as `pinned_durable_ts`.
- `__wt_txn_clear_read_timestamp` is functionally equivalent and is refactored to make handling of durable and read timestamps as similar as possible.

agorrod commented:
@tetsuo-cpp is this ready to merge?
